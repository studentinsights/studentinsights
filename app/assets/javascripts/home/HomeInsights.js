import React from 'react';
import qs from 'query-string';
import Card from '../components/Card';
import Section from '../components/Section';
import GenericLoader from '../components/GenericLoader';
import {apiFetchJson} from '../helpers/apiFetchJson';


// On the home page, show users the answers to their most important questions.
class HomeInsights extends React.Component {
  constructor(props) {
    super(props);
    this.fetchStudentsWithLowGrades = this.fetchStudentsWithLowGrades.bind(this);
    this.renderStudentsWithLowGrades = this.renderStudentsWithLowGrades.bind(this);
  }

  fetchStudentsWithLowGrades() {
    const {educatorId, limit} = this.props;
    const params = educatorId ? {limit, educator_id: educatorId} : {limit};
    const url = `/home/students_with_low_grades_json?${qs.stringify(params)}`;
    return apiFetchJson(url);
  }

  render() {
    return (
      <div className="HomeInsights" style={styles.root}>
        <GenericLoader
          style={styles.card}
          promiseFn={this.fetchStudentsWithLowGrades}
          render={this.renderStudentsWithLowGrades} />
        {this.renderPlaceholder()}
      </div>
    );
  }

  renderStudentsWithLowGrades(json) {
    const props = {
      limit: json.limit,
      totalCount: json.total_count,
      studentsWithLowGrades: json.students_with_low_grades
    };
    return <CheckStudentWithLowGradesView {...props} />;
  }

  renderPlaceholder() {
    return (
      <Card style={styles.card}>
        <div style={styles.placeholderCard}>
          <div>What else would help you support your students?</div>
          <div>Come talk with us about what we should build next!</div>
        </div>
      </Card>
    );
  }
}
HomeInsights.propTypes = {
  educatorId: React.PropTypes.number.isRequired,
  limit: React.PropTypes.number // limit the data returned, not the query itself
};
HomeInsights.defaultProps = {
  limit: 100
};


// Pure UI component, for showing a high school teacher
// which of their students have low grades but haven't been
// discussed in NGE or 10GE.  The intention is that this list of
// students to check in on is immediately actionable.
export class CheckStudentWithLowGradesView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      uiLimit: 4
    };
    this.onMoreStudents = this.onMoreStudents.bind(this);
  }

  onMoreStudents(e) {
    e.preventDefault();
    const {uiLimit} = this.state;
    this.setState({ uiLimit: uiLimit + 4 });
  }

  render() {
    const {studentsWithLowGrades, totalCount} = this.props;
    const {uiLimit} = this.state;
    const truncatedStudentsWithLowGrades = studentsWithLowGrades.slice(0, uiLimit);

    return (
      <div className="CheckStudentWithLowGrades">
        <div style={styles.cardTitle}>NGE and 10GE students to check in on</div>
        <Card style={{border: 'none'}}>
           <div>There {this.renderAreHowManyStudents(totalCount)} in your classes who have a D or an F right now but no one has mentioned in NGE or 10GE for the last month.</div>
          {this.renderList(truncatedStudentsWithLowGrades)}
          {this.renderMore(truncatedStudentsWithLowGrades)}
        </Card>
      </div>
    );
  }

  renderAreHowManyStudents(totalCount) {
    if (totalCount === 0) return <span>are <b>no students</b></span>;
    if (totalCount === 1) return <span>is <b>one student</b></span>;
    return <span>are <b>{totalCount} students</b></span>;
  }

  renderList(truncatedStudentsWithLowGrades) {
    if (truncatedStudentsWithLowGrades.length === 0) return null;
    return (
      <div style={{paddingTop: 10, paddingBottom: 10}}>
        {truncatedStudentsWithLowGrades.map(studentWithLowGrades => {
          const {student, assignments} = studentWithLowGrades;
          return (
            <div key={student.id} style={styles.line}>
              <span><a style={styles.person} href={`/students/${student.id}`}>{student.first_name} {student.last_name}</a></span>
              {this.renderCourseGrades(assignments)}
            </div>
          );
        })}
      </div>
    );
  }

  renderCourseGrades(assignments) {
    return (
      <span>{assignments.map(assignment => {
        const {section} = assignment;
        const hasAnText = (assignment.grade_letter === 'F')
          ? 'has an'
          : 'has a';
        return (
          <span key={assignment.id}>
            <span style={styles.middleText}>{hasAnText} {assignment.grade_letter} in</span>
            <Section
              id={section.id}
              courseDescription={section.course_description}
              sectionNumber={section.section_number} />
          </span>
        );
      })}</span>
    );
  }

  renderMore(truncatedStudentsWithLowGrades) {
    const {totalCount, limit, studentsWithLowGrades} = this.props;

    if (truncatedStudentsWithLowGrades.length !== studentsWithLowGrades.length) {
      return <div><a href="#" onClick={this.onMoreStudents}>See more</a></div>;
    }

    if (studentsWithLowGrades.length < totalCount) {
      return <div>There are {totalCount} students total.  Start with checking in on these first {limit} students.</div>;
    }

    return null;
  }
}

CheckStudentWithLowGradesView.propTypes = {
  limit: React.PropTypes.number.isRequired,
  totalCount: React.PropTypes.number.isRequired,
  studentsWithLowGrades: React.PropTypes.arrayOf(React.PropTypes.shape({
    student: React.PropTypes.shape({
      id: React.PropTypes.number.isRequired,
      first_name: React.PropTypes.string.isRequired,
      last_name: React.PropTypes.string.isRequired,
      grade: React.PropTypes.string.isRequired,
      house: React.PropTypes.string
    }).isRequired,
    assignments: React.PropTypes.arrayOf(React.PropTypes.shape({
      id: React.PropTypes.number.isRequired,
      grade_letter: React.PropTypes.string.isRequired,
      grade_numeric: React.PropTypes.string.isRequired,
      section: React.PropTypes.shape({
        id: React.PropTypes.number.isRequired,
        section_number: React.PropTypes.string.isRequired,
        educators: React.PropTypes.arrayOf(React.PropTypes.object).isRequired
      }).isRequired
    }))
  })).isRequired
};


const styles = {
  root: {
    fontSize: 14
  },
  card: {
    margin: 10,
    marginTop: 20,
    border: '1px solid #ccc',
    borderRadius: 3
  },
  cardTitle: {
    backgroundColor: '#eee',
    padding: 10,
    color: 'black',
    borderBottom: '1px solid #ccc'
  },
  person: {
    fontWeight: 'bold'
  },
  placeholderCard: {
  },
  line: {
    paddingLeft: 10,
    paddingRight: 10,
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap'
  },
  middleText: {
    display: 'inline-block',
    paddingLeft: 5,
    paddingRight: 5
  }
};


export default HomeInsights;