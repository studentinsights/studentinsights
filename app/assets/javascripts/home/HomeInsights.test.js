import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import fetchMock from 'fetch-mock/es5/client';
import HomeInsights, {CheckStudentWithLowGradesView} from './HomeInsights';
import SpecSugar from '../../../../spec/javascripts/support/spec_sugar.jsx';
import studentsWithLowGradesJson from '../../../../spec/javascripts/fixtures/home_students_with_low_grades_json';

function renderIntoEl(element) {
  const el = document.createElement('div');
  ReactDOM.render(element, el);
  return el;
}

function testProps() {
  return {
    educatorId: 9999
  };
}

function pureTestPropsForN(n) {
  return {
    limit: 10,
    totalCount: n,
    studentsWithLowGrades: studentsWithLowGradesJson.students_with_low_grades.slice(0, n)
  };
}

describe('HomeInsights', () => {
  beforeEach(() => {
    fetchMock.restore();
    fetchMock.get('/home/students_with_low_grades_json?educator_id=9999&limit=100', studentsWithLowGradesJson);
  });

  it('renders without crashing', () => {
    const props = testProps();
    const el = document.createElement('div');
    ReactDOM.render(<HomeInsights {...props} />, el);
  });

  SpecSugar.withTestEl('integration tests', container => {
    it('renders everything after fetch', done => {
      const props = testProps();
      const el = container.testEl;
      ReactDOM.render(<HomeInsights {...props} />, el);
      
      setTimeout(() => {
        expect($(el).text()).toContain('There are 7 students in your classes');
        done();
      }, 0);
    });
  });
});

describe('CheckStudentWithLowGradesView', () => {
  it('renders zero, singular and plural states', () => {
    expect($(renderIntoEl(<CheckStudentWithLowGradesView {...pureTestPropsForN(0)} />)).text()).toContain('There are no students');
    expect($(renderIntoEl(<CheckStudentWithLowGradesView {...pureTestPropsForN(1)} />)).text()).toContain('There is one student');
    expect($(renderIntoEl(<CheckStudentWithLowGradesView {...pureTestPropsForN(4)} />)).text()).toContain('There are 4 students');
  });

  it('pure component matches snapshot', () => {
    const props = {
      limit: studentsWithLowGradesJson.limit,
      totalCount: studentsWithLowGradesJson.total_count,
      studentsWithLowGrades: studentsWithLowGradesJson.students_with_low_grades
    };
    const tree = renderer
      .create(<CheckStudentWithLowGradesView {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });


  it('handles when limit reached', () => {
    const props = {
      limit: 3,
      totalCount: 129,
      studentsWithLowGrades: studentsWithLowGradesJson.students_with_low_grades.slice(0, 3)
    };
    const tree = renderer
      .create(<CheckStudentWithLowGradesView {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});