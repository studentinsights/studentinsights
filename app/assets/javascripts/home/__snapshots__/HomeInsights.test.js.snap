// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckStudentWithLowGradesView handles when limit reached 1`] = `
<div
  className="CheckStudentWithLowGrades"
>
  <div
    style={
      Object {
        "backgroundColor": "#eee",
        "borderBottom": "1px solid #ccc",
        "color": "black",
        "padding": 10,
      }
    }
  >
    NGE and 10GE students to check in on
  </div>
  <div
    className="Card "
    style={
      Object {
        "border": "none",
        "borderRadius": 3,
        "padding": 10,
      }
    }
  >
    <div>
      There 
      <span>
        are 
        <b>
          129
           students
        </b>
      </span>
       in your classes who have a D or an F right now but no one has mentioned in NGE or 10GE for the last month.
    </div>
    <div
      style={
        Object {
          "paddingBottom": 10,
          "paddingTop": 10,
        }
      }
    >
      <div
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": 10,
            "paddingRight": 10,
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <span>
          <a
            href="/students/66"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Chip
             
            Duck
          </a>
        </span>
        <span>
          <span>
            <span
              style={
                Object {
                  "display": "inline-block",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              has a
               
              D
               in
            </span>
            <a
              className="Section"
              href="/sections/1"
              style={Object {}}
            >
              BIOLOGY 1 HONORS
               (
              SHS-BIO-TUES
              )
            </a>
          </span>
        </span>
      </div>
      <div
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": 10,
            "paddingRight": 10,
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <span>
          <a
            href="/students/2"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Mari
             
            Kenobi
          </a>
        </span>
        <span>
          <span>
            <span
              style={
                Object {
                  "display": "inline-block",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              has a
               
              D
               in
            </span>
            <a
              className="Section"
              href="/sections/1"
              style={Object {}}
            >
              BIOLOGY 1 HONORS
               (
              SHS-BIO-TUES
              )
            </a>
          </span>
        </span>
      </div>
      <div
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": 10,
            "paddingRight": 10,
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <span>
          <a
            href="/students/69"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Aladdin
             
            Poppins
          </a>
        </span>
        <span>
          <span>
            <span
              style={
                Object {
                  "display": "inline-block",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              has a
               
              D
               in
            </span>
            <a
              className="Section"
              href="/sections/1"
              style={Object {}}
            >
              BIOLOGY 1 HONORS
               (
              SHS-BIO-TUES
              )
            </a>
          </span>
        </span>
      </div>
    </div>
    <div>
      There are 
      129
       students total.  Start with checking in on these first 
      3
       students.
    </div>
  </div>
</div>
`;

exports[`CheckStudentWithLowGradesView pure component matches snapshot 1`] = `
<div
  className="CheckStudentWithLowGrades"
>
  <div
    style={
      Object {
        "backgroundColor": "#eee",
        "borderBottom": "1px solid #ccc",
        "color": "black",
        "padding": 10,
      }
    }
  >
    NGE and 10GE students to check in on
  </div>
  <div
    className="Card "
    style={
      Object {
        "border": "none",
        "borderRadius": 3,
        "padding": 10,
      }
    }
  >
    <div>
      There 
      <span>
        are 
        <b>
          7
           students
        </b>
      </span>
       in your classes who have a D or an F right now but no one has mentioned in NGE or 10GE for the last month.
    </div>
    <div
      style={
        Object {
          "paddingBottom": 10,
          "paddingTop": 10,
        }
      }
    >
      <div
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": 10,
            "paddingRight": 10,
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <span>
          <a
            href="/students/66"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Chip
             
            Duck
          </a>
        </span>
        <span>
          <span>
            <span
              style={
                Object {
                  "display": "inline-block",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              has a
               
              D
               in
            </span>
            <a
              className="Section"
              href="/sections/1"
              style={Object {}}
            >
              BIOLOGY 1 HONORS
               (
              SHS-BIO-TUES
              )
            </a>
          </span>
        </span>
      </div>
      <div
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": 10,
            "paddingRight": 10,
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <span>
          <a
            href="/students/2"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Mari
             
            Kenobi
          </a>
        </span>
        <span>
          <span>
            <span
              style={
                Object {
                  "display": "inline-block",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              has a
               
              D
               in
            </span>
            <a
              className="Section"
              href="/sections/1"
              style={Object {}}
            >
              BIOLOGY 1 HONORS
               (
              SHS-BIO-TUES
              )
            </a>
          </span>
        </span>
      </div>
      <div
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": 10,
            "paddingRight": 10,
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <span>
          <a
            href="/students/69"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Aladdin
             
            Poppins
          </a>
        </span>
        <span>
          <span>
            <span
              style={
                Object {
                  "display": "inline-block",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              has a
               
              D
               in
            </span>
            <a
              className="Section"
              href="/sections/1"
              style={Object {}}
            >
              BIOLOGY 1 HONORS
               (
              SHS-BIO-TUES
              )
            </a>
          </span>
        </span>
      </div>
      <div
        style={
          Object {
            "overflow": "hidden",
            "paddingLeft": 10,
            "paddingRight": 10,
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <span>
          <a
            href="/students/67"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Aladdin
             
            Kenobi
          </a>
        </span>
        <span>
          <span>
            <span
              style={
                Object {
                  "display": "inline-block",
                  "paddingLeft": 5,
                  "paddingRight": 5,
                }
              }
            >
              has a
               
              D
               in
            </span>
            <a
              className="Section"
              href="/sections/1"
              style={Object {}}
            >
              BIOLOGY 1 HONORS
               (
              SHS-BIO-TUES
              )
            </a>
          </span>
        </span>
      </div>
    </div>
    <div>
      <a
        href="#"
        onClick={[Function]}
      >
        See more
      </a>
    </div>
  </div>
</div>
`;
